fish-rust/src/builtins/tests/string_tests.rs

```diff
@@ -0,0 +1,309 @@
+use crate::ffi_tests::add_test;
+
+add_test! {"test_string", || {
+    use crate::wchar::WString;
+    use crate::ffi::parser_t;
+    use crate::ffi;
+    use crate::builtins::string::string;
+    use crate::wchar_ffi::WCharFromFFI;
+    use crate::common::{EscapeStringStyle, escape_string};
+    use crate::wchar::L;
+    use crate::builtins::shared::{STATUS_CMD_ERROR,STATUS_CMD_OK, STATUS_INVALID_ARGS};
+    use crate::future_feature_flags::{feature_test, FeatureFlag};
+    use crate::future_feature_flags::mutable_fish_features;
+
+    // TODO: these should be individual tests, not all in one, port when we can run these with `cargo test`
+    macro_rules! string_test {
+        ($args:expr, $expected_rc:expr, $expected_out:expr) => {
+            let parser: &mut parser_t = unsafe { &mut *parser_t::principal_parser_ffi() };
+
+            let mut streams = ffi::make_test_io_streams_ffi();
+            let mut io = crate::builtins::shared::io_streams_t::new(streams.pin_mut());
+
+            let rc = string(parser, &mut io, $args.as_mut_slice()).expect("string failed");
```

Comment:
This apparently contains a memory leak according to the address sanitiser? All the ffi-io-stuff is very janky and this will be fixed once we can use `IoStream` from Rust (which btw is not usable in anyway since it does not own the `OutputStream`, and you thus cannot call any `append`-functions on it since that requires mutable references)

Reply:
The leak is [from here](https://github.com/fish-shell/fish-shell/pull/9854/files#diff-762b4ab1431326324fde5c8721dced2a973e2d6a00c2e131411ff8c9331943f9R416).

Probably the simplest way to fix this is with a subclass:

struct owning_io_streams_t : io_streams_t {
    string_output_stream_t out_storage;
    null_output_stream_t err_storage;
    owning_io_streams_t() : io_streams_t(out_storage, err_storage) {}
};

and then equip `io_streams_t` with a virtual destructor.

Alternatively since this is very temporary I'd be fine with an ugly hack, e.g. just suppress the leak by adding it to a global vector or something.